hardware/chisel/src/main/scala/core/TensorAlu.scala

```diff
@@ -97,38 +97,330 @@ class AluVector(implicit p: Parameters) extends Module {
+    resetData=false.B, en = true.B)
+  val src_valid_002 = RegNext( src_valid_001, init=false.B)
+  val src_valid_003 = RegNext( src_valid_002, init=false.B)
+  val src_valid_004 = RegNext( src_valid_003, init=false.B)
+
+  val dst_idx_001 = ShiftRegister( index_generator.io.dst_idx, aluDataReadPipeDelay + 1)
+  val src_idx_001 = ShiftRegister( index_generator.io.src_idx, aluDataReadPipeDelay + 1)
+
+  val uop_data_001 = ShiftRegister(io.uop.data, aluDataReadPipeDelay)
+
+  val dst_offset = uop_data_001.bits.u0
+
+  val w = dst_offset.getWidth
+  val u2 = uop_data_001.bits.u2.asTypeOf(UInt(w.W))
+  val s = log2Ceil(p(CoreKey).inpMemDepth)
+  val u1 = uop_data_001.bits.u1.asTypeOf(UInt(w.W))
+  val src_offset = (u2 << s) | u1
+
+  // split registers of stage 2 by data groups
+  //val accRdIdxValid = valid_002 || src_valid_002
+  val accRdIdxValid = valid_001 || src_valid_001
+  for (idx <- 0 until dataSplitFactor) {
+    //io.acc.rd(idx).idx.valid := accRdIdxValid
+    io.acc.rd(idx).idx.valid := RegNext(accRdIdxValid)
+  }
+
+  val new_src_idx_001 = src_idx_001 + src_offset
+  val src_idx_002 = RegNext( new_src_idx_001)
+  val src_idx_003 = RegNext( src_idx_002)
+
+  val new_dst_idx_001 = dst_idx_001 + dst_offset
+  val dst_idx_002 = RegNext( new_dst_idx_001)
+  val dst_idx_003 = RegNext( dst_idx_002)
+  val dst_idx_004 = RegNext( dst_idx_003)
+
+  // split registers of stage 2 by data groups
+  val accRdIdxBits = Mux( src_valid_001 || io.dec.alu_use_imm, new_src_idx_001, new_dst_idx_001)
+  for (idx <- 0 until dataSplitFactor) {
+    io.acc.rd(idx).idx.bits := RegNext(accRdIdxBits)
+    assert( io.acc.rd(idx).data.valid === (valid_003 || src_valid_003))
+  }
+
+  require(io.out.splitWidth == 1 && io.out.splitLength == 1, "-F- Out split write is not supported")
+  val numVecUnits = dataSplitFactor
+  val outData = Wire(io.out.wr(0).bits.data.cloneType)
+  val dataRemapB = Wire(Vec(numVecUnits, io.acc.rd(0).data.bits.cloneType))
+  val dataRemapA = Wire(Vec(numVecUnits, io.acc.rd(0).data.bits.cloneType))
+  // numVecUnits is a pow of 2
+  // split dec bits pipe further if there are many vecUnits
+  val decSplitNb0 =  if (numVecUnits < 8) 1 else 2
+  val decSplit0 = Wire(Vec(decSplitNb0, io.dec.cloneType))
+  for (idx <- 0 until decSplitNb0) {
+    decSplit0(idx) := ShiftRegister(io.dec, if(aluDataReadPipeDelay < 2) 0 else 1)
+  }
+
+  for (idx <- 0 until numVecUnits) {
+    val alu = Module(new AluVector)
+
+    for(aluLenIdx <- 0 until alu.io.acc_b.lenSplit) {
+      for(aluWdtIdx <- 0 until alu.io.acc_b.widthSplit) {
+        val (accGrpIdx, accLenIdx, accWdtIdx) =
+          alu.io.acc_b.reindexDataFromGroup(idx, aluLenIdx, aluWdtIdx)
+        dataRemapB(idx)(aluLenIdx)(aluWdtIdx) :=
+          io.acc.rd(accGrpIdx).data.bits(accLenIdx)(accWdtIdx)
+      }
+    }
+    val save_src = RegNext(dataRemapB(idx))
+    val tensorImm = Wire(new TensorClientData(tensorType = "acc"))
+    tensorImm.data.valid := RegNext(valid_002) //valid_003 split
```

Comment:
maybe creating a binding for `valid_r3` and then assigned it to `tensorImm.data.valid` instead of comment

Reply:
Reusing existing node